### ISSUE TITLE
Rethink binding mode switching

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#2918]
As mentioned in #2917, reversing `fish_vi_mode` is currently a bit weird, and the main docs still mention `fish_vi_key_bindings` as the main way to get vi-mode.

This could be fixed in a variety of ways:

- Set $fish_key_bindings and $__fish_vi_mode in fish_vi_key_bindings and make that the way to switch

- Make a `fish_emacs_mode` function to reverse fish_vi_mode

- Eliminate $__fish_vi_mode, simply check $fish_vi_key_bindings - this means setting fish_key_bindings would suffice

**Comment by krader1961** [comment]
When I started using fish six months ago I was, and still am, confused why there is a distinction between `fish_vi_mode` and `fish_vi_key_bindings`. Not to mention the lack of equivalents for emacs mode and bindings. The challenge before us is how to fix this mess without breaking backward compatibility.

I'm inclined to +1 @faho's third recommendation to eliminate `__fish_vi_mode` (presumably by making it equivalent to `fish_vi_key_bindings`.

**Comment by floam** [comment]
It seems like there should be one base setting that simply indicates, one or the other. They are exclusive and should not clobber each other - I can have a set of emacs-mode keybindings I have set up how I like and also vi. 

It's very confusing right now.

Abstractly, I want to think of it as:

`operating-mode = vi`, or `operating-mode = emacs`. 

**#3** sounds closest to that. I like that.

**Comment by faho** [comment]
>What's important for me is that switching is not destructive. It should switch between two exclusive bind mappings

Here's a case that's a bit tougher to fix:

```fish
set fish_key_bindings fish_default_key_bindings
bind \r 'commandline -f accept-autosuggestion execute'
set fish_key_bindings fish_vi_key_bindings
bind \r # will be the default binding
```

This is why we demand that bindings be placed in fish_user_key_bindings (which is a bit confusing to newcomers), and it also means that switching is destructive to anything you've interactively set. Solving it for simple `bind` statements in config.fish would be easy - set the default bindings before reading it (currently we set it up after the first fish_prompt event IIRC).

But solving it in general would necessitate keeping track of what's a default binding (for the mode) and what's not. At least as far as I can tell.

**Comment by krader1961** [comment]
> What's important for me is that switching is not destructive.

As @faho pointed out that isn't really feasible if you expect the switch to honor arbitrary `bind` commands entered at an interactive prompt. On the other hand if you simply mean that anything in your `fish_user_key_bindings` should be honored then, yes, we agree and that should be the behavior.

Regarding bindings shared between the two modes such as `\cC` (i.e., [ctrl-C]). That is easily handled if we agree with my [proposal](https://github.com/fish-shell/fish-shell/issues/2832#issuecomment-203168544) in issue #2832 that fish specific key bindings be distinct from the bindings for a specific style of editing. Then each editing style simply includes the fish specific bindings before imposing their editing mode specific bindings.

**Comment by faho** [comment]
I've just noticed we have another variable in play here - "$fish_bind_mode" - that contains the current vi-mode (i.e. "default", "insert" etc).

So one possibility is to have the mode indicator use that for determining if it should display anything. We'd just need to erase it in `fish_default_key_bindings` and we could get rid of `$__fish_vi_mode`. This would make it a bit easier to add a non-vi-style binding with modes (not sure such a thing exists, though).

We can also make the key_bindings functions set $fish_key_bindings to their values if it isn't already, i.e.

```fish
if test "$fish_key_bindings" != "fish_vi_key_bindings"
    set fish_key_bindings fish_vi_key_bindings # This triggers the handler, which then calls us again
    return
end
```

in fish_vi_key_bindings. This would make both executing the binding functions _and_ setting $fish_key_bindings "just work". `fish_vi_mode` could be kept as a (deprecated) alias. It's even one of those things that could just scream if they are executed ("This function is deprecated, please just call fish_vi_key_bindings").

**Comment by faho** [comment]
Nope, "fish_bind_mode" won't work as it's also set in default mode (the word "mode" really loses all meaning here, right?). The rest is okay, it just needs to actually `test` the value of $fish_key_bindings.

**Comment by faho** [comment]
I've now implemented this and it seems to work great, only it makes bind.expect fail on my system for some reason.

**Comment by floam** [comment]
This goes for 2.3.0, right? This is a much better default. There's no hard and fast code/feature/behavior freezes here, right?

**Comment by faho** [comment]
Okay, the issue is that my changes (on [my bind-mode branch](https://github.com/faho/fish-shell/tree/bind-mode)) enable the mode prompt where previously it wouldn't be.

I consider that a good change, but it does require changing the `expect_prompt` function in our expect stuff (for the interactive tests).

The closest I've gotten is adding an addition `.?.?.?.?` - I can't seem to get the regexp dialect (or rather the _escaping_ dialect) right to match it better. Any help is appreciated!

**Comment by krader1961** [comment]
Change line 54 of *tests/interactive.expect.rc* to

```
set prompt_pat [list -re "(?:\\r\\n?|^)(?:\\\[.\\\] )?prompt $prompt_counter>(?:$|\\r)"]
```

The triple backslashes are because brackets are special inside TCL double-quoted strings and we need to escape the brackets in the regex: So `\\` + `\[` becomes `\[` when passed to the regex engine.

P.S., I don't understand why the vi-mode prompt doesn't have any coloration since *share/functions/fish_mode_prompt.fish* has the colors hardcoded and I don't see anywhere that the `set_color` function is defined to be no-op.

**Comment by faho** [comment]
Thanks!

>P.S., I don't understand why the vi-mode prompt doesn't have any coloration since share/functions/fish_mode_prompt.fish has the colors hardcoded and I don't see anywhere that the set_color function is defined to be no-op.

I'm assuming it's either $TERM or stdout not being a tty in that context.

**Comment by faho** [comment]
Fixed by merging #2939.
